src/l2c_capability.c

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (C) 2016 Swift Navigation Inc.
+ * Contact: Dmitry Tatarinov <dmitry.tatarinov@exafore.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+#include <stdio.h>
+#include <ch.h>
+#include <assert.h>
+
+#include <libswiftnav/common.h>
+#include <libswiftnav/nav_msg.h>
+#include <libswiftnav/logging.h>
+
+#include "cfs/cfs-coffee.h"
+#include "cfs/cfs.h"
+
+#include "l2c_capability.h"
+
+static MUTEX_DECL(l2c_capability_upd_mutex);
+
+/** gps_l2c_sv_capability 
+  L2C capability info containter
+  Each bit responses for a SV:
+  Bit 0 - SV 1, Bit 31 - SV 32.
+  bit set - L2C stream available for the SV
+  bit cleared - L2C stream not available
+  By default assume all SV can broadcast L2C stream */
+static u32 gps_l2c_sv_capability = 0xffffffff;
+
+/** \defgroup l2c_capability L2C_Capability
+ * Functions used in l2c capability
+ * \{ */
+
+/** Update gps_l2c_sv_capability. And write it to Flash memory if
+ * previous value of gps_l2c_sv_capability is different.
+ * \param l2c_cpbl New value of L2C capability 
+ *
+ */
+void l2c_capability_update(u32 l2c_cpbl)
+{
+  if (l2c_cpbl == gps_l2c_sv_capability)
+    return;
+
+  chMtxLock(&l2c_capability_upd_mutex);
+
+  gps_l2c_sv_capability = l2c_cpbl;
+
+  int fd = cfs_open(L2C_FILE, CFS_WRITE);
+  if (fd != -1) {
+    if(cfs_write(fd,
```

Comment:
Flash writes from the decoder thread scare me a little bit. I don't know what the maximum write time is, but if we do this for several channels back to back it might cause the nav bit FIFOs to overrun. Right now the FIFOs are 32 nav bits per channel, which can be as little as 64ms for some encodings.

I think doing the writes directly can be ok as long as we're sure timing won't be violated, but a better option might be to set a flag and periodically write to flash from a background thread when necessary (assuming the write can be done in a way that doesn't block the decode thread).